### PR TITLE
Add BYOI subcommand to import public OS image.

### PR DESCRIPTION
**Pull request by HanFa**
### What this PR does / why we need it

This PR adds a subcommand for Tanzu called `image` as a temporary workaround to enable bring-your-own-image. This subcommand imports OSImage from a public image url to user's account / compartment, and then patches the TKr with private OSImage resource. The patched TKr can then be used to create a management cluster with `--additional-manifests`.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

manually invocation of the subcommand

### Release note

```release-note
Add BYOI subcommand to import public OS image.
```

### CODE REVIEW

**Comment by randomvariable** [comment]
Let's move this under the cluster plugin, so the end result is something like

`tanzu cluster os-image <infrastructure-provider>`, e.g.

`tanzu cluster os-image oracle`

**Comment by randomvariable** [comment]
Secondly, move the calls to Oracle SDK to `tkg/oracle`

The CLI should be a thin wrapper around package functions.

File: cmd/cli/plugin/cluster/osimage_oracle.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/aunum/log"
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8sjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	utilyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	"github.com/vmware-tanzu/tanzu-framework/cli/core/pkg/carvelhelpers"
```
Comment (anujc25) [inline cmd/cli/plugin/cluster/osimage_oracle.go:21]: Plugins should not use `cli/core` module as a dependency. Please use `tkg/carvelhelpers` package within `tkg` module. You might need to do changes done under `cli/core/pkg/carvelhelpers/registry.go` to `tkg/carvelhelpers/registry.go`

**Comment by HanFa** [comment]
thanks @anujc25 for reviewing! I've changed its dependency to use `"github.com/vmware-tanzu/tanzu-framework/tkg/carvelhelpers"` instead. Could you ptal?

File: cmd/cli/plugin/cluster/osimage_oracle.go (hunk 1 above)
Comment (HanFa) [reply cmd/cli/plugin/cluster/osimage_oracle.go]: fixed

File: cmd/cli/plugin/cluster/osimage_oracle.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+					"id":          &imageId,
+					"compartment": &compartment,
+					"region":      region,
+				}},
+		},
+	}
+	if err = writeToFile(tkr, getTKRPath()); err != nil {
+		return err
+	}
+	if err = writeToFile(osImage, getOSImage()); err != nil {
```
Comment (tenczar) [inline cmd/cli/plugin/cluster/osimage_oracle.go:131]: If this fails, should we clean up the tkr written in the previous block?

**Review by tenczar — COMMENTED** [review]
Few comments about file handling. Looks good in general

File: cmd/cli/plugin/cluster/osimage_oracle.go (hunk 2 above)
Comment (randomvariable) [reply cmd/cli/plugin/cluster/osimage_oracle.go:131]: I think it's ok to leave it so the user can at least manually do something with it as an escape hatch.

File: cmd/cli/plugin/cluster/osimage_oracle.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+func writeToFile(object runtime.Object, path string) error {
+	serializer := k8sjson.NewSerializerWithOptions(
+		k8sjson.DefaultMetaFactory, nil, nil,
+		k8sjson.SerializerOptions{
+			Yaml:   true,
+			Pretty: true,
+			Strict: true,
+		},
+	)
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, os.ModePerm)
```
[the remark this answers is not in the document]
Comment (HanFa) [reply cmd/cli/plugin/cluster/osimage_oracle.go:62]: good catch! fixed

File: cmd/cli/plugin/cluster/osimage_oracle.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	if err = serializer.Encode(object, file); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getTKRFromManifest reads the TanzuKubernetesRelease from its yaml manifest given path
+func getTKRFromManifest(path string) (*v1alpha3.TanzuKubernetesRelease, error) {
+	var tkr v1alpha3.TanzuKubernetesRelease
+	tkrFile, err := os.Open(path)
```
[the remark this answers is not in the document]
Comment (HanFa) [reply cmd/cli/plugin/cluster/osimage_oracle.go:76]: good catch! fixed

**Pull request merged**